# Level 1

# TASK 3 – Cleaning Data

## Objective
Demonstrate professional-level data cleaning skills by taking a deliberately messy dataset and systematically transforming it into a clean, analysis-ready dataset.

## Tech Stack
- Python
- pandas
- NumPy
- Jupyter Notebook

## Project Overview
In this project, we will inspect a messy dataset, identify data quality issues, handle missing values and duplicates, standardise inconsistent data, detect and treat outliers, correct data types, and save the final cleaned dataset for analysis.


In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("messy_retail_customers.csv")
df.head()

,customer_id,customer_name,gender,age,purchase_date,purchase_amount,city,payment_method
0,C001,Alice,F,24,2025-01-05,1250.50,Hyderabad,Credit Card
1,C002,Bob,male,31,05/01/2025,890,hyderabad,UPI
2,C003,Charlie,M,45,2025/01/07,"₹1,450",Bangalore,Cash
3,C004,Diana,Female,NaN,Jan 08 2025,2100.00,Bengaluru,upi
4,C005,Ethan,m,29,2025-01-09,750.25,Chennai,Credit Card


In [3]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])
print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

Number of rows: 33
Number of columns: 8

Column names:
['customer_id', 'customer_name', 'gender', 'age', 'purchase_date', 'purchase_amount', 'city', 'payment_method']

Data types:
customer_id        object
customer_name      object
gender             object
age                object
purchase_date      object
purchase_amount    object
city               object
payment_method     object
dtype: object


In [4]:
print("Missing values in each column:")
print(df.isnull().sum())
print("\nTotal duplicate rows:", df.duplicated().sum())

Missing values in each column:
customer_id        0
customer_name      0
gender             1
age                2
purchase_date      0
purchase_amount    1
city               0
payment_method     1
dtype: int64

Total duplicate rows: 1


In [5]:
print("Gender values:")
print(df["gender"].value_counts(dropna=False))
print("\nCity values:")
print(df["city"].value_counts(dropna=False))
print("\nPayment method values:")
print(df["payment_method"].value_counts(dropna=False))
print("\nAge values:")
print(df["age"].tolist())
print("\nPurchase amount values:")
print(df["purchase_amount"].tolist())

Gender values:
gender
M         7
Female    7
Male      5
female    4
F         3
male      3
m         2
NaN       1
FEMALE    1
Name: count, dtype: int64

City values:
city
Hyderabad     6
Bangalore     4
Chennai       4
Pune          4
Mumbai        4
Delhi         4
Bengaluru     3
hyderabad     1
chennai       1
Mumbai        1
chennai       1
Name: count, dtype: int64

Payment method values:
payment_method
UPI            12
Credit Card     9
Cash            8
upi             1
NaN             1
credit card     1
cash            1
Name: count, dtype: int64

Age values:
['24', '31', '45', nan, '29', '36', '41', '27', '33', '30', '999', '26', '39', '28', '34', '25', '43', '32', nan, '29', '37', '48', '22', '35', '31', '40', '52', '27', '38', '44', '29', 'thirty', '26']

Purchase amount values:
['1250.50', '890', '₹1,450', '2100.00', '750.25', '1,250', '980.75', '500', nan, '1750.00', '1200', '-450', '2300', '670.50', '1,050', '890.00', '15000', '999.99', '810', '720.00', '1100', '95

In [6]:
duplicate_count = df.duplicated().sum()
df = df.drop_duplicates()
print("Duplicate rows removed:", duplicate_count)
print("Duplicate rows remaining:", df.duplicated().sum())
print("Rows remaining:", len(df))

Duplicate rows removed: 1
Duplicate rows remaining: 0
Rows remaining: 32


## Missing Data Handling

The dataset contains missing values in `gender`, `age`, `purchase_amount`, and `payment_method`.

- **Gender:** Missing gender values will be filled using the **mode**, because gender is a categorical variable and the most frequent category provides a reasonable replacement.
- **Age:** Missing age values will be filled using the **median**. Median is preferred because the dataset contains extreme and invalid age values, making it more robust than the mean.
- **Purchase Amount:** Missing purchase amounts will be filled using the **median** because transaction amounts can contain extreme values, and the median is less affected by outliers.
- **Payment Method:** Missing payment method values will be filled using the **mode**, because this is a categorical variable and the most common payment method is an appropriate replacement.

These choices are made to preserve as much data as possible while reducing the influence of extreme values.

In [7]:
df["age"] = pd.to_numeric(df["age"], errors="coerce")
print(df["age"])

0      24.0
1      31.0
2      45.0
3       NaN
4      29.0
5      36.0
6      41.0
7      27.0
8      33.0
9      30.0
10    999.0
11     26.0
12     39.0
13     28.0
14     34.0
15     25.0
16     43.0
17     32.0
18      NaN
19     29.0
20     37.0
21     48.0
22     22.0
23     35.0
24     31.0
25     40.0
26     52.0
27     27.0
28     38.0
29     44.0
31      NaN
32     26.0
Name: age, dtype: float64


In [8]:
median_age = df["age"].median()
print("Median age:", median_age)

Median age: 33.0


In [9]:
df["age"] = df["age"].fillna(median_age)
print("Missing age values after imputation:", df["age"].isnull().sum())

Missing age values after imputation: 0


In [10]:
df["gender"] = df["gender"].replace({
    "M": "Male",
    "m": "Male",
    "male": "Male",
    "F": "Female",
    "female": "Female",
    "FEMALE": "Female"})
print(df["gender"].value_counts(dropna=False))

gender
Male      16
Female    15
NaN        1
Name: count, dtype: int64


In [11]:
gender_mode = df["gender"].mode()[0]
print("Gender mode:", gender_mode)

Gender mode: Male


In [12]:
df["gender"] = df["gender"].fillna(gender_mode)
print("Missing gender values after imputation:", df["gender"].isnull().sum())

Missing gender values after imputation: 0


In [13]:
df["payment_method"] = df["payment_method"].replace({
    "upi": "UPI",
    "credit card": "Credit Card",
    "cash": "Cash"})
print(df["payment_method"].value_counts(dropna=False))

payment_method
UPI            13
Credit Card     9
Cash            9
NaN             1
Name: count, dtype: int64


In [14]:
payment_mode = df["payment_method"].mode()[0]
print("Payment method mode:", payment_mode)

Payment method mode: UPI


In [15]:
df["payment_method"] = df["payment_method"].fillna(payment_mode)
print("Missing payment method values after imputation:", df["payment_method"].isnull().sum())

Missing payment method values after imputation: 0


In [16]:
df["purchase_amount"] = (
    df["purchase_amount"]
    .astype("string")
    .str.replace("₹", "", regex=False)
    .str.replace("$", "", regex=False)
    .str.replace(",", "", regex=False)
)
df["purchase_amount"] = pd.to_numeric(
    df["purchase_amount"],
    errors="coerce"
)
print(df["purchase_amount"])
print("\nData type:", df["purchase_amount"].dtype)

0      1250.5
1       890.0
2      1450.0
3      2100.0
4      750.25
5      1250.0
6      980.75
7       500.0
8        <NA>
9      1750.0
10     1200.0
11     -450.0
12     2300.0
13      670.5
14     1050.0
15      890.0
16    15000.0
17     999.99
18      810.0
19      720.0
20     1100.0
21      950.3
22      600.0
23     1250.0
24      980.0
25     1300.0
26      875.8
27     1020.0
28     1450.0
29      760.0
31     1100.0
32      850.5
Name: purchase_amount, dtype: Float64

Data type: Float64


In [17]:
Q1_age = df["age"].quantile(0.25)
Q3_age = df["age"].quantile(0.75)
IQR_age = Q3_age - Q1_age

lower_age = Q1_age - 1.5 * IQR_age
upper_age = Q3_age + 1.5 * IQR_age

age_outliers = df[
    (df["age"] < lower_age) |
    (df["age"] > upper_age)
]
print("Q1:", Q1_age)
print("Q3:", Q3_age)
print("IQR:", IQR_age)
print("Lower limit:", lower_age)
print("Upper limit:", upper_age)
print("\nAge outliers:")
print(age_outliers[["customer_id", "age"]])

Q1: 28.75
Q3: 39.25
IQR: 10.5
Lower limit: 13.0
Upper limit: 55.0

Age outliers:
   customer_id    age
10        C011  999.0


## Outlier Treatment – Age

The IQR method identified one outlier in the `age` column:

- Customer `C011` has an age of `999`.
- The calculated upper IQR limit is `55`.
- Therefore, `999` is an extreme and unrealistic value.

### Decision
The value `999` will not be retained because it is clearly invalid for an age.

Instead of deleting the complete customer record, the invalid age will be replaced with a missing value and then imputed using the **median age (33)**.

This approach preserves the customer's other valid information while correcting the invalid age entry.

In [18]:
df.loc[df["age"] == 999, "age"] = np.nan
df["age"] = df["age"].fillna(median_age)
print("Age value for C011:", df.loc[df["customer_id"] == "C011", "age"].iloc[0])

Age value for C011: 33.0


In [19]:
Q1_amount = df["purchase_amount"].quantile(0.25)
Q3_amount = df["purchase_amount"].quantile(0.75)
IQR_amount = Q3_amount - Q1_amount
lower_amount = Q1_amount - 1.5 * IQR_amount
upper_amount = Q3_amount + 1.5 * IQR_amount

amount_outliers = df[
    (df["purchase_amount"] < lower_amount) |
    (df["purchase_amount"] > upper_amount)
]

print("Q1:", Q1_amount)
print("Q3:", Q3_amount)
print("IQR:", IQR_amount)
print("Lower limit:", lower_amount)
print("Upper limit:", upper_amount)
print("\nPurchase amount outliers:")
print(amount_outliers[["customer_id", "purchase_amount"]])

Q1: 830.25
Q3: 1250.25
IQR: 420.0
Lower limit: 200.25
Upper limit: 1880.25

Purchase amount outliers:
   customer_id  purchase_amount
3         C004           2100.0
11        C012           -450.0
12        C013           2300.0
16        C017          15000.0


## Outlier Treatment – Purchase Amount

The IQR method identified four potential outliers in the `purchase_amount` column. The calculated acceptable range was from 200.25 to 1880.25.

- **C004 – 2100:** Retained because it is a plausible positive purchase amount.
- **C013 – 2300:** Retained because it is a plausible positive purchase amount.
- **C012 – -450:** Treated as invalid because a negative purchase amount is not reasonable for this dataset.
- **C017 – 15000:** Treated as an extreme anomaly because it is substantially higher than the other transaction amounts.

The two invalid/anomalous values will be replaced with missing values and imputed using the median purchase amount. This preserves the customer records while correcting unreasonable transaction values.

In [20]:
df.loc[df["purchase_amount"] < 0, "purchase_amount"] = np.nan
df.loc[df["purchase_amount"] == 15000, "purchase_amount"] = np.nan

median_amount = df["purchase_amount"].median()
print("Median purchase amount:", median_amount)

Median purchase amount: 999.99


In [21]:
df["purchase_amount"] = df["purchase_amount"].fillna(median_amount)
print(
    "Missing purchase amount values after imputation:",
    df["purchase_amount"].isnull().sum()
)

Missing purchase amount values after imputation: 0


In [22]:
df["city"] = df["city"].str.strip().str.title()
df["city"] = df["city"].replace({
    "Bengaluru": "Bangalore"
})

print(df["city"].value_counts())

city
Hyderabad    7
Bangalore    7
Chennai      5
Mumbai       5
Delhi        4
Pune         4
Name: count, dtype: int64


In [23]:
original_dates = pd.read_csv("messy_retail_customers.csv")["purchase_date"]

In [24]:
df["purchase_date"] = pd.to_datetime(
    original_dates,
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

print(df["purchase_date"].head(10))

print("\nMissing dates after correction:",
      df["purchase_date"].isnull().sum())

print("Data type:", df["purchase_date"].dtype)

0   2025-01-05
1   2025-01-05
2   2025-01-07
3   2025-01-08
4   2025-01-09
5   2025-01-10
6   2025-01-11
7   2025-01-12
8   2025-01-13
9   2025-01-14
Name: purchase_date, dtype: datetime64[ns]

Missing dates after correction: 0
Data type: datetime64[ns]


In [25]:
print(df.dtypes)

customer_id                object
customer_name              object
gender                     object
age                       float64
purchase_date      datetime64[ns]
purchase_amount           Float64
city                       object
payment_method             object
dtype: object


In [26]:
text_columns = [
    "customer_id",
    "customer_name",
    "gender",
    "city",
    "payment_method"
]

for col in text_columns:
    df[col] = df[col].astype("string")

df["age"] = df["age"].astype(float)
df["purchase_amount"] = df["purchase_amount"].astype(float)

print(df.dtypes)

customer_id        string[python]
customer_name      string[python]
gender             string[python]
age                       float64
purchase_date      datetime64[ns]
purchase_amount           float64
city               string[python]
payment_method     string[python]
dtype: object


In [27]:
# Before vs. After cleaning summary

original_df = pd.read_csv("messy_retail_customers.csv")

dtype_columns_to_check = {
    "age": pd.api.types.is_numeric_dtype,
    "purchase_date": pd.api.types.is_datetime64_any_dtype,
    "purchase_amount": pd.api.types.is_numeric_dtype
}

before_dtype_issues = sum(
    not check(original_df[column].dtype)
    for column, check in dtype_columns_to_check.items()
)

# Count data type issues after cleaning
after_dtype_issues = sum(
    not check(df[column].dtype)
    for column, check in dtype_columns_to_check.items()
)

before_nulls = original_df.isnull().sum().sum()
before_duplicates = original_df.duplicated().sum()
before_rows = len(original_df)

after_nulls = df.isnull().sum().sum()
after_duplicates = df.duplicated().sum()
after_rows = len(df)

summary = pd.DataFrame({
    "Metric": [
        "Total Null Values",
        "Duplicate Rows",
        "Row Count",
        "Dtype Issues"
    ],
    "Before Cleaning": [
        before_nulls,
        before_duplicates,
        before_rows,
        before_dtype_issues
    ],
    "After Cleaning": [
        after_nulls,
        after_duplicates,
        after_rows,
        after_dtype_issues
    ]
})

summary

,Metric,Before Cleaning,After Cleaning
0,Total Null Values,5,0
1,Duplicate Rows,1,0
2,Row Count,33,32
3,Dtype Issues,3,0


In [28]:
print("Missing values:")
print(df.isnull().sum())

print("\nDuplicate rows:", df.duplicated().sum())

print("\nData types:")
print(df.dtypes)

print("\nFinal number of rows:", len(df))
print("Final number of columns:", len(df.columns))

Missing values:
customer_id        0
customer_name      0
gender             0
age                0
purchase_date      0
purchase_amount    0
city               0
payment_method     0
dtype: int64

Duplicate rows: 0

Data types:
customer_id        string[python]
customer_name      string[python]
gender             string[python]
age                       float64
purchase_date      datetime64[ns]
purchase_amount           float64
city               string[python]
payment_method     string[python]
dtype: object

Final number of rows: 32
Final number of columns: 8


In [29]:
output_file = "cleaned_retail_customers.csv"
df.to_csv(output_file, index=False)
print("Cleaned dataset saved successfully as:", output_file)

Cleaned dataset saved successfully as: cleaned_retail_customers.csv


# Conclusion

The messy retail customer dataset was successfully transformed into a clean, analysis-ready dataset.

The cleaning process included:

- Identification and removal of duplicate records.
- Handling missing values using appropriate imputation strategies.
- Standardisation of gender, city, payment method, and date formats.
- Correction of invalid and inconsistent data values.
- Detection and treatment of outliers using the IQR method.
- Correction of data types for accurate analysis.
- Final validation to confirm that the cleaned dataset contains no missing values or duplicate rows.

The final dataset contains **32 rows and 8 columns** and is ready for further analysis and visualization.

The cleaned dataset was saved as `cleaned_retail_customers.csv`.